src/octoprint/plugins/gcodeviewer/static/js/viewer/reader.js

```diff
@@ -77,34 +80,116 @@ GCODE.gCodeReader = (function () {
         return tmpModel;
     };
 
-    var prepareLinesIndex = function (m) {
-        percentageTree = undefined;
-
-        for (var l = 0; l < m.length; l++) {
-            if (m[l] === undefined) continue;
-            for (var i = 0; i < m[l].length; i++) {
-                var percentage = m[l][i].percentage;
-                var value = {layer: l, cmd: i};
-                if (!percentageTree) {
-                    percentageTree = new AVLTree({key: percentage, value: value}, "key");
+    var searchInPercentageTree = function (key) {
+        function searchInLayers(lower, upper, key) {
+            while (lower < upper) {
+                var middle = Math.floor((lower + upper) / 2);
+
+                if (
+                    rendererModel[middle][0].percentage <= key &&
+                    (!rendererModel[middle + 1] ||
+                        rendererModel[middle + 1][0].percentage > key)
+                )
+                    return middle;
+
+                if (rendererModel[middle][0].percentage > key) {
+                    upper = middle - 1;
                 } else {
-                    percentageTree.add({key: percentage, value: value});
+                    lower = middle + 1;
                 }
             }
+
+            return lower;
         }
-    };
 
-    var searchInPercentageTree = function (key) {
-        if (percentageTree === undefined) {
-            return undefined;
+        function searchInCmds(layer, lower, upper, key) {
+            while (lower < upper) {
+                var middle = Math.floor((lower + upper) / 2);
+
+                if (rendererModel[layer][middle].percentage == key) return middle;
+
+                if (rendererModel[layer][middle].percentage > key) {
+                    upper = middle - 1;
+                } else {
+                    lower = middle + 1;
+                }
+            }
+            return lower;
         }
 
-        var elements = percentageTree.findBest(key);
-        if (elements.length === 0) {
-            return undefined;
+        if (rendererModel === undefined) return undefined;
+
+        // this happens when the print is stopped.
+        // just return last position to keep the last
+        // position on screen.
+        if (key == null) return {layer: cacheLastLayer, cmd: cacheLastCmd};
+
+        var bestLayer = undefined;
+        var bestCmd = undefined;
+
+        // check if we are within cacheLookAhead distance of our last position
+        if (cacheLastLayer !== undefined) {
+            if (
```

Comment:
This needs a comment explaining the meaning of the expression. Something like

```js
// command likely in the cached layer
```

I also can't help but think that there's way too much code repetition in this cached layer handling.

Reply:
Before I try to fix the code repetition by adding function calls: 

A binary search is O(log n), so the number of loop repetition of a full search is limited. We have like max 30000 layers and lets say 100000 commands in a layer. That means a full search is max 15 + 23 iterations of the binary search loop. 

How many loops equals 1 function call in cost? In C, it would be quite a few. (flushing registers to stack, fetching function code, restoring registers from cache after the call, ...).

We may just choose to drop the caching feature altogether and just do a full lookup each time. It would simplify the code considerably. I believe the current code is a speedup but I am less sure if I start adding function calls to that cache lookup code.

Reply:
Probably a good idea because frankly, just off the top of my head I wouldn't be able to say how much the function calls here weigh vs the caching gains.

I'd lean towards "a lot" myself though. Note that I didn't necessarily say that this needed to be consolidated into a function, it just felt so utterly repetitive that I was wondering if there was *something* that could be done to make it more readable because frankly right now it takes a lot of mental energy to process.